### PR TITLE
Major changes to exec, change steps, add step and delete step

### PR DESCRIPTION
**Pull request by Nfsaavedra**
- Allows to go backwards outside of proofs with `exec`
- Support imports in `exec`
- Fix `delete_step`, `add_step` and `change_steps`
- Implement deleting, adding and changing steps outside of proofs

### CODE REVIEW

**Review by pcarrott — CHANGES_REQUESTED** [review]
The `in_proof` flag and `current_goals` are still in `CoqFile`, even though they are no longer used there. All proof handling is done in `ProofFile`, so these fields should be extracted to there.

File: coqpyt/coq/context.py [hunk 1]
```diff
@@ -15,11 +15,15 @@ def __init__(
...
         self.__init_coq_version(coqtop)
-        self.terms = {} if terms is None else terms
-        self.last_term: Optional[Term] = None
+        # NOTE: We use a stack for each term because of the following case:
+        # A) File A imports a file B with term C
+        # b) File A defines a new term C
+        self.__terms: Dict[str, List[Term]] = {} if terms is None else terms
+        self.__last_terms: List[Tuple[str, Term]] = []
         self.__segments = SegmentStack()
         self.__anonymous_id: Optional[int] = None
```
[the remark this answers is not in the document]
Comment (Nfsaavedra) [reply coqpyt/coq/context.py:28]: Yeah, makes sense, forgot to do it.

File: coqpyt/coq/proof_file.py [hunk 2]
```diff
@@ -376,69 +419,57 @@ def __is_proof_term(self, step: Step):
...
         self.__aux_file.append(self.prev_step.text)
         self.__check_program()
-        self.__check_proof_step(self.prev_step, goals)
+        if self.in_proof or self.__is_end_proof(self.prev_step):
+            self.__check_proof_step(self.prev_step, goals)
 
     def __backward_step(self):
         self.__aux_file.truncate(self.curr_step.text)
-        self.__check_proof_step(self.curr_step)
+        if len(self.open_proofs) > 0 or self.__is_end_proof(self.curr_step):
```
[the remark this answers is not in the document]
Comment (pcarrott) [reply coqpyt/coq/proof_file.py:450]: I think I get it. It's because we still need to remove the proof from the `__open_proofs`, even if the step was already taken (i.e., `in_proof` is false). In that case, I think the length of `__open_proofs` should be checked in `__forward_step`, as well. I think it makes sense and it's a good idea to keep the functions as similar as possible. Do you agree?

File: coqpyt/coq/base_file.py [hunk 3]
```diff
@@ -260,93 +256,173 @@ def _make_change(self, change_function, *args):
...
-            text = "".join(lines)
+        with open(self.__path, "w") as f:
             f.write(text)
 
+    def __delete_update_ast(self, step_index: int):
+        deleted_step = self.steps[step_index]
+        for step in self.steps[step_index + 1 :]:
+            remove_chars = deleted_step.ast.range.end.character
+            if deleted_step.ast.range.end.line == deleted_step.ast.range.start.line:
+                remove_chars -= deleted_step.ast.range.start.character - 1
```
[the remark this answers is not in the document]
Comment (Nfsaavedra) [reply coqpyt/coq/base_file.py:313]: Yes, I think you are partially right, it's not because of the `\n` tho, it's because of a white space.

This is tested on `TestProofSimpleFileChanges`. Namely, when removing the reflexivity:
```
Example test1: 1 + 1 = 2. reflexivity. Qed.
```
The reflexivity step starts on character 26 and ends on 38, and that does not consider the white space that I think that we also remove. And that's why I added the -1. Because that way, the remove_chars will be 13 instead of 12.

We have to test it, but I think your change is correct. However, when doing that change we have to check if the step_index is 0 or not, and if it is, use (0,0) as a replacement of `prev_step.ast.range.end`.

File: coqpyt/coq/base_file.py [hunk 4]
```diff
@@ -260,93 +256,173 @@ def _make_change(self, change_function, *args):
...
+
+            if step.ast.range.start.line == deleted_step.ast.range.end.line:
+                step.ast.range.start.character -= remove_chars
+            if step.ast.range.end.line == deleted_step.ast.range.end.line:
+                step.ast.range.end.character -= remove_chars
+
+            step.ast.range.start.line -= deleted_step.text.count("\n")
+            step.ast.range.end.line -= deleted_step.text.count("\n")
+
+    def __add_update_ast(self, previous_step_index: int, step_text: str) -> Step:
```
[the remark this answers is not in the document]
Comment (Nfsaavedra) [reply coqpyt/coq/base_file.py:323]: I'm not totally sure about what changes are you proposing, but the reasoning seems right :+1:

File: coqpyt/coq/proof_file.py [hunk 5]
```diff
@@ -484,98 +550,117 @@ def exec(self, nsteps=1) -> List[Step]:
...
+        steps = self.steps_taken - previous_step_index - 1
+        offset = 1 if previous_step_index + 1 < self.steps_taken else 0
+
+        optional = self.__find_step(self.steps[previous_step_index].ast.range)
+        self._make_change(self._add_step, previous_step_index, step_text)
+
+        # The step was not processed yet
+        if self.steps_taken <= previous_step_index + 1:
+            super().exec(steps + offset)
+            return
```
[the remark this answers is not in the document]
Comment (Nfsaavedra) [reply coqpyt/coq/proof_file.py:592]: Yap, you right, I moved the if and forgot to remove the exec.

File: coqpyt/coq/proof_file.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Nfsaavedra) [reply coqpyt/coq/proof_file.py:450]: Actually there were two reasons:
- The first reason is the one you mentioned
- The second reason is because of performance. `in_proof` can take about 0.5 seconds if the goals were not already obtained.

In `__forward_step`, the reason why I didn't use `__open_proofs` is because of the steps that define the terms. When you create a Theorem, you still don't have `__open_proofs`. We could do the reverse of what we do in the backward steps with the `__is_end_proof` and use the `__is_proof_term`, but the `__is_proof_term` does not have guarantees that a proof was generated. For that reason, the only option that I found was to use `in_proof`.

**Comment by Nfsaavedra** [comment]
For now, I think that we should [skip](https://pytest-with-eric.com/pytest-best-practices/pytest-skip-test/#Unconditional-Skipping) the `TestProofListNotation` and probably add a FIXME comment with a link to issue #24.

File: coqpyt/coq/base_file.py [hunk 6]
```diff
@@ -356,35 +432,61 @@ def _add_step(
...
-                self._delete_step(change.step_index, validate_file=False)
+                self.__delete_step_text(change.step_index)
+                self.__delete_update_ast(change.step_index)
+                self.steps.pop(change.step_index)
                 offset_steps -= 1
             else:
                 raise NotImplementedError(f"Unknown change: {change}")
 
+        self.__update_steps()
         if len(self.steps) != previous_steps_size + offset_steps or not self.is_valid:
```
[the remark this answers is not in the document]
Comment (pcarrott) [reply coqpyt/coq/base_file.py:438]: Indeed, the offset check is need. I added tests and comments explaining why.

File: coqpyt/tests/proof_file/expected/list_notation.yml [hunk 7]
```diff
@@ -20,7 +20,6 @@ proofs:
             line: 4
             character: 7
         context:
-          # FIXME issue #24
```
Comment (Nfsaavedra) [inline coqpyt/tests/proof_file/expected/list_notation.yml:23]: Why did you remove this?

File: coqpyt/coq/proof_file.py [hunk 8]
```diff
@@ -425,30 +443,56 @@ def __is_proof_term(self, step: Step):
...
+        # Ignore segment delimiters because it affects Program handling
+        if self.__is_segment_delimiter(self.curr_step):
+            return
+        # Check if Program definition or proof step
+        is_program = self.__check_program(
+            self.current_goals, prev_goals=goals, undo=True
+        )
+        if not is_program and len(self.open_proofs) > 0:
+            self.__check_proof_step(self.curr_step, goals, undo=True)
```
Comment (Nfsaavedra) [inline coqpyt/coq/proof_file.py:495]: I think we can merge these two functions. We just need a new function, for instance, called __step, that receives the goals, a step and undo.

File: coqpyt/tests/proof_file/test_proof_file.py [hunk 9]
```diff
@@ -156,26 +158,6 @@ def test_nested_proofs(self):
...
-        assert len(proof_file.open_proofs) == 0
-
-        # Re-open proofs
-        proof_file.exec(-2)
-        assert len(proof_file.proofs) == 2
-        assert len(proof_file.open_proofs) == 2
-
-        proof_file.exec(2)
-        assert len(proof_file.proofs) == 4
-        assert len(proof_file.open_proofs) == 0
```
Comment (Nfsaavedra) [inline coqpyt/tests/proof_file/test_proof_file.py:140]: Why did you remove this?

Comment (pcarrott) [reply coqpyt/tests/proof_file/test_proof_file.py]: It was extracted to a new class `TestProofChangeNestedProofs` in `test_changes.py`. That was the only test in `test_proof_file.py` with `add_step`/`delete_step`/`change_steps`, which should only be tested in `test_changes.py`.

File: coqpyt/tests/proof_file/expected/list_notation.yml (hunk 7 above)
Comment (pcarrott) [reply coqpyt/tests/proof_file/expected/list_notation.yml:23]: A FIXME was added to the actual test making this comment redundant.
